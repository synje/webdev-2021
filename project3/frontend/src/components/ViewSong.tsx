import React from 'react'
import { Box } from '@chakra-ui/layout'
import { Button, ModalOverlay, ModalHeader, ModalContent, ModalCloseButton, ModalBody, ButtonGroup, FormLabel } from "@chakra-ui/react"
import { gql, useMutation } from '@apollo/client'

/**
 * Mutation-query to VM, deletes song from the database with the ID of the song the user is viewing
 */
export const DELETE_SONG = gql`
    mutation DeleteSong($songID: ID!){
        deleteSong(songID: $songID){
            _id
        }
    }
    `

/**
 * Exports ViewSong component to component DBList.tsx through Modal (pop-up box from Chakra)
 * Contains details about the song viewed, a button for deleting the song, and functionality for deleting a song
 * 
 * @param param Data about a song on the form {param: {title: "Example", duration: "2.2", ...(etc) }}
 * @returns A modal-view of the song-details (pop-up-box)
 */
function ViewSong(param: any) {
    const title = param.param.title

    /**
     * Submits songID to delete-mutation-query to delete song from database, refreshes window
     * @param paramVariables Data about ID of song the user wants to delete
     */
    const submit = (paramVariables: any) => {
        const clickedSongID = paramVariables.variables.songID
        deleteSong({ variables: { songID: clickedSongID } })
        refreshWindow()
    }

    function refreshWindow() {
        window.location.reload()
    }

    /**
     * Function to mutate database, takes in a mutation-query
     */
    const [deleteSong, { error, loading, data }] = useMutation(DELETE_SONG)
    if (loading) return <p>Submitting...</p>;
    if (error) return <p>Submission error :(( - {error.message}</p>;

    return (
        <>
            <ModalOverlay />
            <ModalContent>
                <ModalHeader>SONG VIEW:</ModalHeader>
                <ModalCloseButton />
                <ModalBody>
                    <Box>
                        <Box
                            mt="1"
                            fontWeight="semibold"
                            as="h4"
                            lineHeight="tight"
                            isTruncated
                        > <Box as="span">
                                Title: {title}
                            </Box>
                        </Box>
                        <Box>
                            <Box as="span">
                                Artist: {param.param.artist}
                            </Box>
                        </Box>
                        <Box>
                            <Box as="span">
                                Duration: {param.param.duration} min
                            </Box>
                        </Box>
                        <Box>
                            <Box as="span">
                                Album: {param.param.album}
                            </Box>
                        </Box>
                        <Box>
                            <Box as="span">
                                Year released: {param.param.year}
                            </Box>
                        </Box>
                    </Box>
                    <Box display="flex" justifyContent="flex-end" mb="10px" mt="10px">
                        <Box w="100%" display="flex" justifyContent="space-between" alignItems="center">
                            <FormLabel htmlFor="deleteSong">Deleting will refresh the page:</FormLabel>
                            <Button data-cy="delete" name="deleteSong" onClick={() => {
                                submit({ variables: { songID: param.param._id } })
                            }}
                                as="span" key={param.param._id} bg="firebrick" color="white"
                            >Delete song
                            </Button>
                        </Box>
                    </Box>
                </ModalBody>
            </ModalContent >
        </>
    )
}

export default ViewSong;