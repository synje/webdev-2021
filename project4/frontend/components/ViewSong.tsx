import React, { useState } from 'react'
import { gql, useMutation } from '@apollo/client'
import { Button, Title } from 'react-native-paper'
import { View, StyleSheet, Text, Platform, Dimensions } from 'react-native'

/**
 * Mutation-query to VM, deletes song from the database with the ID of the song the user is viewing
 */
export const DELETE_SONG = gql`
    mutation DeleteSong($songID: ID!){
        deleteSong(songID: $songID){
            _id
        }
    }
`

/**
 * Exports ViewSong-component to component DBList.tsx
 * Contains details about the song viewed, a button for deleting the song, and functionality for deleting a song
 * 
 * @param param Data about a song on the form {param: {title: "Example", duration: "2.2", ...(etc) }}
 * @returns A modal-view of the song-details (pop-up-view)
 */
function ViewSong(param: any) {

    /** State of disabled on delete-button */
    const [disabledState, setDisabledState] = useState(Boolean);

    /**
     * Submits songID to delete-mutation-query to delete song from database, calls successfulDelete(title)
     * @param paramVariables Data about song the user wants to delete
     */
    const submit = (paramVariables: any) => {
        const clickedSongID = paramVariables.variables.songID
        deleteSong({ variables: { songID: clickedSongID } })
        successfulDelete(paramVariables.variables.title)
    }

    /**
     * Shows an alert on successful delete, disables delete-button
     * @param title 
     */
    function successfulDelete(title: any) {
        alert(`Song "${title}" was successfully deleted!`)
        setDisabledState(true)
    }


    /**
     * Function to mutate database, takes in a mutation-query
     */
    const [deleteSong, { error, loading, data }] = useMutation(DELETE_SONG)
    if (loading) return <Text style={{ marginTop: 10 }}>Submitting...</Text>;
    if (error) return <Text style={{ marginTop: 10 }}>Submission error: {error.message}</Text>


    return (
        <>
            <View>
                <View style={styles.headerBox}>
                    <View style={styles.modalText}>
                        <Text>
                            Title:<Title style={styles.headerText}>{param.param.title}</Title>
                        </Text>
                    </View>
                </View>
                <View style={styles.headerBox}>
                    <View style={styles.modalText}>
                        <Text>
                            Artist:<Title style={styles.headerText}>{param.param.artist}</Title>
                        </Text>
                    </View>
                </View>
                <View style={styles.headerBox}>
                    <View style={styles.modalText}>
                        <Text>
                            Duration:<Title style={styles.headerText}>{param.param.duration} min</Title>
                        </Text>
                    </View>
                </View>
                <View style={styles.headerBox}>
                    <View style={styles.modalText}>
                        <Text>
                            Album:<Title style={styles.headerText}>{param.param.album}</Title>
                        </Text>
                    </View>
                </View>
                <View style={styles.headerBox}>
                    <View style={styles.modalText}>
                        <Text>
                            Year released:<Title style={styles.headerText}>{param.param.year}</Title>
                        </Text>
                    </View>
                </View>
                <View style={styles.headerBoxButton}>
                    <View style={styles.modalText}>
                        <View>
                            <Title style={styles.headerText}>Delete is permanent:</Title>
                            <Button data-cy="delete" style={[styles.deleteButtonStyle, disabledState && styles.disabledButton]} labelStyle={styles.deleteButtonText}
                                onPress={() => {
                                    submit({ variables: { songID: param.param._id, title: param.param.title } })
                                }}
                                key={param.param._id}
                                disabled={disabledState}
                            >Delete song</Button>
                        </View>
                    </View>
                </View>
            </View>
        </>
    )
}

export default ViewSong;

const screenHeight = Dimensions.get('screen').height
const styles = StyleSheet.create({
    modalText: {
        marginBottom: 5,
        width: "100%",
    },
    headerBox: {
        display: "flex",
        alignItems: "center",
        borderBottomWidth: 2,
        borderBottomColor: "#A9A9A9",
        marginTop: 15,
    },
    headerBoxButton: {
        display: "flex",
        alignItems: "center",
        marginTop: 15,
    },
    headerText: {
        color: "black",
        marginLeft: 5,
        fontSize: 20,
    },
    deleteButtonStyle: {
        backgroundColor: "#8B0000",
        width: "100%",
        height: 35,
        marginTop: 8,
        marginBottom: 15,
    },
    deleteButtonText: {
        color: "white",
        fontSize: 12,
    },
    disabledButton: {
        backgroundColor: "#808080",
    },
    total: {
        position: 'absolute'
    },
})