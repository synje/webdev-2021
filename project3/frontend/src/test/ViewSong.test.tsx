import ViewSong, { DELETE_SONG } from "../components/ViewSong"
import { MockedProvider } from "@apollo/react-testing";
import TestRenderer, { ReactTestRendererJSON } from 'react-test-renderer';


const mockDeleteSong = {
    request: {
        query: DELETE_SONG,
        variables: {
            _id: "1234",
            title: "RED-SUN (021)",
            duration: 3.09,
            album: "THE PARK IN THE NIGHT part three",
            year_released: 2019,
            aritst: {
                name1: 'GWSN',
            },
        },
    },
    error: new Error('Error in request'),
};

let property = { _id: "", title: "", duration: 0, album: "", year: 0, artist: [""] }
function getView(_id: string, title: string, duration: number, album: string, year: number, artist: Array<any>) {
    let artistArray: any[] = []
    artist.map((name: any) => {
        artistArray.push(name.name1)
        if (artist[0].name2) {
            artistArray.push(", ", name.name2)
        }
    })
    property = { _id: _id, title: title, duration: duration, album: album, year: year, artist: artistArray }
}
getView("698gjd7494hfjdk", "tittel", 0.12, "album", 2012, [{name1: "Artist1"}])


it('ViewSong loades', async () => {
    const component = await TestRenderer.create(
        <MockedProvider mocks={[mockDeleteSong as any]}>
            {getView("698gjd7494hfjdk", "tittel", 0.12, "album", 2012, [{name1: "Artist1"}])}
            <ViewSong property = {property}/>
        </MockedProvider>
    )

    const tree = component.toJSON() as ReactTestRendererJSON
    if(tree != null){
        expect(tree.children).toContain('Loading...')
    
    }
    }
)

test('Snapshot ViewSong', async () =>{
    const tree = TestRenderer.create(
    <MockedProvider mocks={[mockDeleteSong as any]}>
         {getView("698gjd7494hfjdk", "tittel", 0.12, "album", 2012, [{name1: "Artist1"}])}
        <ViewSong property={property} />
    </MockedProvider>)
    expect(tree).toMatchSnapshot()

})