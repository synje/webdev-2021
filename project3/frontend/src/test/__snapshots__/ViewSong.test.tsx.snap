// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot ViewSong 1`] = `null`;
